### PR TITLE
 Moving a dependency on pkg/api/v1/pod and switching to component-helper

### PR DESCRIPTION
**Pull request by yuzhiquan**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
(cleanup): remove direct import k8s.io/kubernetes/pkg/api/v1/pod to pkg/scheduler/util
There is other refer usage, so does not remove older function.
**Which issue(s) this PR fixes**:

Ref #89930

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by alculquicondor** [comment]
What is the other import that you are referring to?

Could we not overload the `pkg/scheduler/util` package and have some kind of `podhelpers` or `apihelpers` package?

**Comment by damemi** [comment]
@alculquicondor from a quick look, outside of scheduler these helpers are used in:
`GetPodPriority`:
* pkg/api/v1/pod/util_test.go
* pkg/kubelet/eviction/helpers.go

`UpdatePodCondition`:
* pkg/api/v1/pod/util_test.go
* pkg/controller/daemon/daemon_controller_test.go
* pkg/controller/statefulset/stateful_set_{control,utils}_test.go
* pkg/controller/util/node/controller_utils.go

`GetPodConditionFromList` is called in a couple other places but for the scheduler it's only used within the other functions. `GetPodCondition` is only used in pkg/api/v1/pod and pkg/kubelet/status.

By enumerating these, would it be reasonable for some of these packages to call into the scheduler helpers, given that we ultimately might want to externalize those helpers anyway?

**Comment by alculquicondor** [comment]
> By enumerating these, would it be reasonable for some of these packages to call into the scheduler helpers, given that we ultimately might want to externalize those helpers anyway?

That sounds quite reasonable. But also that's another argument for making a package that is as narrow as possible. In this case, we are trying to expose helpers around core APIs.

**Comment by yuzhiquan** [comment]
Function in `pkg/scheduler/util/utils.go` is  refer to pod operation, so remove all of them to new pkg, like `pkg/apis/scheduling/podhelper` ?

**Comment by alculquicondor** [comment]
or perhaps podutil, so that functions can be something like: `podutil.GetAffinityTerms`

**Comment by alculquicondor** [comment]
cc @Huang-Wei

**Comment by alculquicondor** [comment]
/hold

I prefer that we fix #89092 first, so that we can carry out patches more easily to older releases.

**Comment by yuzhiquan** [comment]
> /hold
> 
> I prefer that we fix #89092 first, so that we can carry out patches more easily to older releases.

Ok, but seems like will import an new pkg

**Review by damemi — CHANGES_REQUESTED** [review]
I see that the only other place `GetPodPriority` is used is from `pkg/kubelet/eviction/helpers.go` in the priority function. Could you update this package to call the new scheduler util function directly? (and thus be able to remove the original `GetPodPriority`)

As well as copying the TestGetPodPriority function from https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/pod/util_test.go#L835

**Comment by damemi** [comment]
Also, can you update the title of this PR, it is actually removing a dependency on `pkg/api/v1/pod` and switching to pkg/scheduler/util

**Comment by yuzhiquan** [comment]
@damemi respond

**Review by damemi — APPROVED** [review]
/lgtm
@ahg-g this is one of the ones where we're moving helpers to `pkg/scheduler/util`. The only other place these functions were called was in a kubelet helper, so I think in the interest of collecting our scheduler-specific helpers in one spot this one is good, then we can work on externalizing them all in staging later. wdyt?

**Comment by yuzhiquan** [comment]
rebase PR

**Review by damemi — APPROVED** [review]
/lgtm
reapplying after rebase

File: pkg/scheduler/util/utils.go
Comment (ahg-g) [inline pkg/scheduler/util/utils.go:138]: @damemi I am worried about this function, it is not trivial, and so I am afraid we will diverge from the canonical way of updating pod status. As we discussed before, what we want is to get those functions into staging first.

Comment (damemi) [reply pkg/scheduler/util/utils.go:138]: @ahg-g you are right, I thought this function had been *moved* to our utils (along with the others), but I didn't realize it was actually *copied*. Looking at this now I see that several of these functions are duplicated, which I agree is not what we want. Sorry about that.

Where in staging do you think it would be appropriate to move these functions? Maybe apimachinery?

/lgtm cancel

**Comment by damemi** [comment]
/remove-lifecycle rotten
These helpers could be good candidates to go to `component-helpers` once that staging repo is initialized in https://github.com/kubernetes/kubernetes/pull/92507

**Comment by ehashman** [comment]
This PR has needed a rebase for quite some time and has not received any updates in 7+ months. I'm going to close it, but please feel free to reopen if you're ready to pick this back up again!

/close

**Comment by damemi** [comment]
@yuzhiquan are you still interested in this PR? If not I can pick up your commits and move some of them into component-helpers like mentioned above

**Comment by yuzhiquan** [comment]
> @yuzhiquan are you still interested in this PR? If not I can pick up your commits and move some of them into component-helpers like mentioned above

Sorry for lately reply, i thought this pr was already discarded. i will look into this and move on. @damemi

**Comment by yuzhiquan** [comment]
/hold 
wait for @damemi review.

**Comment by damemi** [comment]
Hi @yuzhiquan, thanks for updating this (and sorry it got delayed so long)

One of our goals with component-helpers is to reduce code duplication, so we don't just want to copy functions there but actually move them (and update the usage in other areas to refer to the moved location)

Could you update this PR to remove the old functions and find other callers and update those to import from component-helpers as well?

**Comment by yuzhiquan** [comment]
@damemi 
Sounds good.But in that way, `UpdatePodCondition` should not in component-helpers/scheduling pkg, these functiones called by many other pkg, like kubelet | controller...

**Comment by ahg-g** [comment]
> @damemi
> Sounds good.But in that way, `UpdatePodCondition` should not in component-helpers/scheduling pkg, these functiones called by many other pkg, like kubelet | controller...

kubelet/controller etc. can link to any of those component-helpers pkgs, but the helpers are placed in the component that has the strongest affiliation

**Comment by yuzhiquan** [comment]
Update as comment.
/assign @deads2k @damemi

File: pkg/api/v1/pod/util.go
Comment (damemi) [inline pkg/api/v1/pod/util.go:300]: it might be worth it to move this function to corev1helpers as well, since it is just a wrapper for `GetPodCondition`

**Review by damemi — COMMENTED** [review]
this looks great, moving the functions to component-helpers is a good move. One comment around the `GetPodReadyCondition` function that perhaps an api approver could comment on but that's it from me

File: pkg/api/v1/pod/util.go
Comment (yuzhiquan) [reply pkg/api/v1/pod/util.go]: Done

**Comment by damemi** [comment]
/lgtm
from scheduler side. @liggitt could you ptal for API?

**Comment by liggitt** [comment]
The extraction looks good, but this seems like a pod/node-centric helper, since it is not only used by the scheduler but by the kubelet, controllers, storage plugins, and e2e tests. Since the pod API is primarily under sig-node, would `k8s.io/component-helpers/node/corev1/helpers.go` (with node reviewers/approvers) be a better destination?

**Comment by yuzhiquan** [comment]
> The extraction looks good, but this seems like a pod/node-centric helper, since it is not only used by the scheduler but by the kubelet, controllers, storage plugins, and e2e tests. Since the pod API is primarily under sig-node, would `k8s.io/component-helpers/node/corev1/helpers.go` (with node reviewers/approvers) be a better destination?

@damemi WDYT?

**Comment by damemi** [comment]
Yeah, I agree with @liggitt that it probably fits better under pod/node than scheduling.

>  Since the pod API is primarily under sig-node, would `k8s.io/component-helpers/node/corev1/helpers.go` (with node reviewers/approvers) be a better destination?

sgtm. And we already have `k8s.io/component-helpers/node` with node approvers, so a new `corev1` dir should inherit that

**Comment by yuzhiquan** [comment]
I have moved pod util functions to `k8s.io/component-helpers/node/corev1/helpers.go`, PLZ have a look?
@damemi @liggitt

**Comment by yuzhiquan** [comment]
ping @damemi @liggitt

**Comment by damemi** [comment]
Hi @yuzhiquan, k8s is currently in code freeze for the next release and merges won't be able to happen until at least April 8 for [code thaw](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.21#tldr). Sorry for the delay in this, but I have approved as much as I am able to, and we will need the other Owners to sign off now.

**Comment by yuzhiquan** [comment]
> Hi @yuzhiquan, k8s is currently in code freeze for the next release and merges won't be able to happen until at least April 8 for [code thaw](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.21#tldr). Sorry for the delay in this, but I have approved as much as I am able to, and we will need the other Owners to sign off now.

@damemi Sure, thanks for the help.

**Comment by alculquicondor** [comment]
This needs to be rebased

**Comment by yuzhiquan** [comment]
> This needs to be rebased

@alculquicondor Rebase done.

File: pkg/api/v1/pod/util.go
Comment (yuzhiquan) [reply pkg/api/v1/pod/util.go:301]: This function called by many pkg, may be another pr?

Comment (alculquicondor) [reply pkg/api/v1/pod/util.go:301]: sg

scheduler changes lgtm

**Comment by liggitt** [comment]
API bits lgtm. Assign me once this has approval node packages and I'll tag for the API package.

**Comment by yuzhiquan** [comment]
Ping @derekwaynecarr @lavalamp @soltysh

**Comment by yuzhiquan** [comment]
Ping @derekwaynecarr @lavalamp @soltysh

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by soltysh** [comment]
> /hold
> 
> I prefer that we fix #89092 first, so that we can carry out patches more easily to ol

This landed 
/hold cancel

**Comment by yuzhiquan** [comment]
> > /hold
> 
> > 
> 
> > I prefer that we fix #89092 first, so that we can carry out patches more easily to ol
> 
> 
> 
> This landed 
> 
> /hold cancel

Thanks for review.
And ping @liggitt @mrunalp

**Comment by yuzhiquan** [comment]
ping @alculquicondor @damemi @ehashman @soltysh for review and approve, this pr last for many days

File: pkg/kubelet/status/status_manager.go
Comment (alculquicondor) [inline pkg/kubelet/status/status_manager.go:799]: use `GetPodCondition`

File: pkg/kubelet/status/generate.go
Comment (alculquicondor) [inline pkg/kubelet/status/generate.go:116]: It looks like `GeneratePodReadyCondition` could be changed to accept a `PodStatus` instead of 3 separate arguments. With that, we can make `GetPodConditionFromList` private.

File: staging/src/k8s.io/component-helpers/node/corev1/helpers.go
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/node/corev1/helpers.go:26]: Add tests

Comment (yuzhiquan) [reply staging/src/k8s.io/component-helpers/node/corev1/helpers.go:26]: This function lack of test case, should we add in this pr or another?

Comment (ahg-g) [reply staging/src/k8s.io/component-helpers/node/corev1/helpers.go:26]: add in this one.

Comment (yuzhiquan) [reply staging/src/k8s.io/component-helpers/node/corev1/helpers.go:26]: Add unit test case for GetPodCondition

**Comment by yuzhiquan** [comment]
ping @lavalamp for approve

**Comment by yuzhiquan** [comment]
ping @lavalamp for approve

**Comment by yuzhiquan** [comment]
ping @alculquicondor for approve

**Comment by alculquicondor** [comment]
You already have @damemi's approval for sig scheduling

**Comment by yuzhiquan** [comment]
ping @ehashman for kubelet part approve

**Comment by yuzhiquan** [comment]
ping @liggitt for api part approve.

**Pull request closed without merging**
